### PR TITLE
GitHub issues

### LINKED ISSUE #37: Make timers in the Machine screen update in real time
Without server connection.

### PR DESCRIPTION
**Pull request by jpitlor**
Resolves issues #36 and #30. When I get more time, I'll fix #37 next

### CODE REVIEW

**Comment by jpitlor** [comment]
I realize that I meant to do more testing before committing because the PR doesn't actually remove leading 0s. I'll fix that soon

**Comment by jpitlor** [comment]
Fixed and ready for someone else to make sure I didn't break everything

**Comment by vidia** [comment]
I assume we both updated gradle at different times. Yours is probably old in comparison if I had to guess.

File: app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java [hunk 1]
```diff
@@ -13,6 +13,8 @@
     private String status;
     private String time;
 
+    // TODO: Call constructor when creating machines so stripping leading 0s doesn't
```
Comment (vidia) [inline app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java]: I think these are created automatically with Retrofit, so either "setName()" or the constructor can be trusted to be called every time. I am not completely sure which though.

File: app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java [hunk 2]
```diff
@@ -21,7 +23,10 @@ public Machine(String name, String type, String status, String time) {
     }
 
     public String getName() {
-        return name;
+        // Get rid of leading 0s
+        if (name.matches(".*0[1-9][0-9].*")) return name.replaceFirst("0", "");
+        else if (name.matches(".*0{2}[1-9].*")) return name.replaceAll("0", "");
```
Comment (vidia) [inline app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java]: Can you do a loop with a "charAt()"? Idk what the regex performance is like in Java.

Comment (vidia) [reply app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java]: This might be best done on the server actually. Not that I really want to touch the deployed server though.

File: app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java [hunk 3]
```diff
@@ -37,18 +40,27 @@
...
     private LocationAdapter adapter;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+
+        if (!getIntent().getBooleanExtra("forceMainMenu", false)) {
+            String lastRoom = SharedPrefsHelper.getSharedPrefs(this)
+                    .getString("lastRoom", "hi, jeff");
+            if (!lastRoom.equals("hi, jeff")) {
```
Comment (vidia) [inline app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java]: You can make the alternative be `null` to do this check.

File: app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java [hunk 4]
```diff
@@ -37,18 +40,27 @@
     @Bind(R.id.location_activity_toolbar) Toolbar toolbar;
     @Bind(R.id.progressBar) ProgressBar mLoadingProgressBar;
 
-
     private LocationAdapter adapter;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+
```
Comment (vidia) [inline app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java:48]: Do you need to make a check here to make sure that this is a cold open of the app? Also clear the lastRoom pref once you do this.

File: app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java [hunk 5]
```diff
@@ -37,18 +40,27 @@
...
+        if (!getIntent().getBooleanExtra("forceMainMenu", false)) {
+            String lastRoom = SharedPrefsHelper.getSharedPrefs(this)
+                    .getString("lastRoom", "hi, jeff");
+            if (!lastRoom.equals("hi, jeff")) {
+                Intent intent = new Intent(this, MachineActivity.class);
+                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
```
Comment (vidia) [inline app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java]: Why did you add the flag?

File: app/src/main/java/xyz/jhughes/laundry/activities/MachineActivity.java [hunk 6]
```diff
@@ -145,6 +146,10 @@ public boolean onCreateOptionsMenu(Menu menu) {
     @Override
     public boolean onOptionsItemSelected(MenuItem item) {
         switch (item.getItemId()) {
+            case android.R.id.home:
+                    Intent i = getParentActivityIntent().putExtra("forceMainMenu", true);
+                    navigateUpTo(i);
```
Comment (vidia) [inline app/src/main/java/xyz/jhughes/laundry/activities/MachineActivity.java]: What is this doing?

File: app/src/main/java/xyz/jhughes/laundry/adapters/MachineAdapter.java [hunk 7]
```diff
@@ -81,13 +84,25 @@ private void machineHelper(Machine m, Boolean dryers, String options) {
...
         // create a new view
         View v = LayoutInflater.from(parent.getContext())
                 .inflate(R.layout.cardview_machine, parent, false);
         // set the view's size, margins, paddings and layout parameters
 
+        //updateTimes.scheduleAtFixedRate(new TimerTask() {
+            //@Override
+            //public void run() {
+//                for (int i = 0; i < parent.getChildCount(); i++) {
+//                    View child = parent.getChildAt(i);
```
Comment (vidia) [inline app/src/main/java/xyz/jhughes/laundry/adapters/MachineAdapter.java]: If this code isnt working can you remove it?

**Review by vidia — DISMISSED** [review]
I think this will let you see my code review. 

Most of the comments are just questions, but there are some optimizations that could be made.

File: app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java (hunk 1 above)
Comment (jpitlor) [reply app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java]: Is the time that the machine object persists short enough that it doesn't matter? I can't remember under what circumstances the client makes another request to the server, so it could be. I was just thinking of how I could optimize

File: app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java (hunk 2 above)
Comment (jpitlor) [reply app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java]: I could. But I agree, a server change would be better, and would eliminate the above concern. But isn't the server Tylers, not ours? Or did he give the code/ownership to us?

File: app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java (hunk 3 above)
Comment (jpitlor) [reply app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java]: yeah, I probably should. I'll make that change

File: app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java (hunk 4 above)
Comment (jpitlor) [reply app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java:48]: See, I was having a ton of trouble with this particular feature since the lifecycle says that the exact same methods are called when the user does a cold open and when they navigate back following the event of the phone running low on memory, so unless you have a suggestion, i couldn't find a different good way to check that.

Why clear lastRoom? Do they not want to come back to that room when they close the app a second time?

File: app/src/main/java/xyz/jhughes/laundry/activities/MachineActivity.java (hunk 6 above)
Comment (jpitlor) [reply app/src/main/java/xyz/jhughes/laundry/activities/MachineActivity.java]: Simulating the back button

File: app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java (hunk 1 above)
Comment (vidia) [reply app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java]: When you refresh the list or open a new screen basically. getName is still called more often than the constructor though. 

Its a really minor optimization. Probably doesnt matter too much.

File: app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java (hunk 5 above)
Comment (jpitlor) [reply app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java]: Ah, that was from a different attempt that didn't work. I forgot to take it away. I'll do that

File: app/src/main/java/xyz/jhughes/laundry/adapters/MachineAdapter.java (hunk 7 above)
Comment (jpitlor) [reply app/src/main/java/xyz/jhughes/laundry/adapters/MachineAdapter.java]: Of course not! We need to make the repo as confusing as possible! (yes)

File: app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java (hunk 2 above)
Comment (vidia) [reply app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java]: The server is on AWS with Docker and Redis and all that. We even have reports that come back with status occasionally. 

We have access to the AWS, we just need to actually make the change, and I am not entirely sure how to be honest. But we do have the required control.

File: app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java (hunk 4 above)
Comment (vidia) [reply app/src/main/java/xyz/jhughes/laundry/activities/LocationActivity.java:48]: I thought there was a difference between onNewIntent and onCreate() being called. OnCreate being on the first open and onNewIntent or something being called when a currently opened activity gets a new call. 

So we would only want it in onCreate() I guess. Maybe what you have is correct, hmm.

File: app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java (hunk 2 above)
Comment (jpitlor) [reply app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java]: gotcha. Well I'd be happy to help implement it server-side when we investigate that

Comment (vidia) [reply app/src/main/java/xyz/jhughes/laundry/LaundryParser/Machine.java]: It still works client-side which is good. I just want to be able to trust the data we are given as "clean" and not have to tweak it. 

I'm comfortable taking this change anyway though in the meantime. 

Though I think this will create an issue with sorting the machines. Do you see that happening? 

If so, the sorting code needs to change, or machines need sequential IDs.

File: app/src/main/java/xyz/jhughes/laundry/activities/MachineActivity.java (hunk 6 above)
Comment (vidia) [reply app/src/main/java/xyz/jhughes/laundry/activities/MachineActivity.java]: The back button works for me in the current app? Is there a bug I don't know about?

File: app/src/main/java/xyz/jhughes/laundry/adapters/MachineAdapter.java (hunk 7 above)
Comment (vidia) [reply app/src/main/java/xyz/jhughes/laundry/adapters/MachineAdapter.java]: Lol. Something something security through obscurity/job security.

**Comment by vidia** [comment]
This should be good once I tweak how lastRoom is managed in SPrefs.

**Comment by vidia** [comment]
Removed lastRoom from the app entirely in #50

**Comment by vidia** [comment]
I added the following 

```
@Override
    protected void onPause() {
        super.onPause();
        SharedPreferences.Editor e = SharedPrefsHelper.getSharedPrefs(this).edit();
        e.putString("lastScreenViewed", "LocationList");
        e.apply();
    }
```

to all the activities. If the list is showing then it is set to "LocationList" if its a machine view its set to the room name. 

So you just need to change the key and check for "LocationList" and if so, just skip the opening new things. 

See #50 most recent commit

**Review by hughesjeff — APPROVED** [review]
